rcl/src/rcl/domain_id.c

```diff
@@ -40,10 +41,14 @@ rcl_get_default_domain_id(size_t * domain_id)
       get_env_error_str);
     return RCL_RET_ERROR;
   }
-  if (ros_domain_id) {
+  if (ros_domain_id && strcmp(ros_domain_id, "") != 0) {
```

Comment:
can a `else {*domain_id = 0u;}` be added instead of all the PRs addressing rmw implementations?

Reply:
What if they don't want to use `0` as the default?

Reply:
> What if they don't want to use 0 as the default?

We used to force that in the past, and a lot of the rmw implementations aren't handling the default correctly (thus the other PRs are needed).
I don't mind if we want  to change this.

Reply:
After this PR, [rcl_node_get_domain_id](https://github.com/ros2/rcl/blob/b9cfc243bf018091b5c5bcb5585d19591daf4ae6/rcl/include/rcl/node.h#L371) started to return 0 when the environment variable is unset.

The purpose of that API was to return the actual domain id that the node was using (see [here](https://github.com/ros2/rcl/blob/acd167e1734c2c15196fb92f76bff1559260eb1a/rcl/src/rcl/node.c#L488)), but the "actual" domain id is now erroneous.

IMO, the behavior before this PR was correct, i.e. using 0 as the default domain id if the environment variable is not set.
I really don't see any value in letting the rmw implementation define the actual domain id, and adding a layer in rmw to query the default domain id actually used seems unnecessary.

@hidmic @wjwwood what do you think?

Reply:
> IMO, the behavior before this PR was correct, i.e. using 0 as the default domain id if the environment variable is not set.

As I've stated elsewhere ([here](https://github.com/ros2/rmw_connext/pull/427)?) I disagree.

> I really don't see any value in letting the rmw implementation define the actual domain id, and adding a layer in rmw to query the default domain id actually used seems unnecessary.

I wonder what's the use case for `rcl_get_node_domain_id`. 

> After this PR, rcl_node_get_domain_id started to return 0 when the environment variable is unset.

Hmm, how's that possible? I would expect it to be `RCL_DEFAULT_DOMAIN_ID` w/o further changes.

Reply:
> Hmm, how's that possible? I would expect it to be RCL_DEFAULT_DOMAIN_ID w/o further changes.

Ok, perhaps I misunderstood. `rcl_get_default_domain_id()` behaves as expected.

> I wonder what's the use case for rcl_get_node_domain_id.

If we really need it, yeah, it'll require new `rmw`  API. Having said that, nothing prevents `rcl` from forcing `0` though:

```c
  domain_id = node->impl->options.domain_id;
  if (RCL_DEFAULT_DOMAIN_ID == domain_id) {
    if (RCL_RET_OK != rcl_get_default_domain_id(&domain_id)) {
      goto fail;
    }
+   if (RCL_DEFAULT_DOMAIN_ID == domain_id) {
+     domain_id = 0u;
+   }
  }
  RCUTILS_LOG_DEBUG_NAMED(ROS_PACKAGE_NAME, "Using domain ID of '%zu'", domain_id);
```

We've backported this to Foxy already, so that may be the only solution there.

Reply:
> I wonder what's the use case for rcl_get_node_domain_id.

I think it's good to be able to introspect the domain id, and if it returns SIZE_MAX when it's actually using 0 it might be confusing (I know of some people that were confused).
There was also a log saying `domain_id = -1` which was confusing (I have to double check from where that log came from, the log [here](https://github.com/ros2/rcl/blob/b9cfc243bf018091b5c5bcb5585d19591daf4ae6/rcl/src/rcl/node.c#L263) seem to have the correct format string)).

> Hmm, how's that possible? I would expect it to be RCL_DEFAULT_DOMAIN_ID w/o further changes.

Ups, edited. It started to return SIZE_MAX.

> As I've stated elsewhere (here?) I disagree.

I think you're referring to this:

> IMHO the issue is that both domain ID defaults are conflated. rcl has a default to decide whether it has to check the environment or not, and rmw has a default with a sentinel value. If no ID is passed to rcl and the environment does no provide one either, the rmw default should be honored.
> We can discuss the rmw default though. It could be 0, sure, if we think no rmw can benefit from knowing if that ID is the default and not a user-defined one that happens to have the same value.

There wasn't a RMW_DEFAULT_DOMAIN_ID in the past (see [blame](https://github.com/ros2/rmw/blame/master/rmw/include/rmw/domain_id.h#L19)), that was moved from rcl.
Maybe I should have duplicated the domain id in rcl init options instead of moving the definition of the default, as the default was ONLY checked in `rcl`. What is the value of a default in `rmw`?

> If we really need it, yeah, it'll require new rmw API. Having said that, nothing prevents rcl from forcing 0 though:

That patch looks ok.

Reply:
> There wasn't a RMW_DEFAULT_DOMAIN_ID in the past (see blame), that was moved from rcl.

Interesting. Fun chain of events then. 

> Maybe I should have duplicated the domain id in rcl init options instead of moving the definition of the default, as the default was ONLY checked in rcl. What is the value of a default in rmw?

Well, an `rmw` implementation is expected to provide a default domain ID via [`rmw_init_options_init()`](https://github.com/ros2/rmw/blob/4a1914a3b8c36a0bb59677cdbab9a21fb0da8ddd/rmw/include/rmw/init_options.h#L74). From an API perspective, the sole value I see with having a sentinel ID is to defer picking an actual default (allowing, say, an implementation to pick up a default from a network-wide pool when it is actually needed and not when options are initialized).

> IMHO the issue is that both domain ID defaults are conflated. rcl has a default to decide whether it has to check the environment or not, and rmw has a default with a sentinel value. If no ID is passed to rcl and the environment does no provide one either, the rmw default should be honored.

And that remains true. I see `rcl_init_options_t` using `rmw_init_options_t` domain ID [directly](https://github.com/ros2/rcl/blob/b9cfc243bf018091b5c5bcb5585d19591daf4ae6/rcl/src/rcl/init_options.c#L140). And then implicitly relies on it being `RMW_DEFAULT_DOMAIN_ID == RCL_DEFAULT_DOMAIN_ID` to know whether it has to check the environment or not.

There's absolutely no guarantee that `rmw_init_options_init()` will use `RMW_DEFAULT_DOMAIN_ID` as opposed to `42` or `5`. Which would then break environment checks.

Reply:
> That patch looks ok.

I think that:

- For Foxy, we don't have a choice. `rcl_get_default_domain_id()` has to force a 0 or `rcl_node_get_domain_id()` breaks as you said.
- For Galactic, I think we need `rmw` API **and** fix those assumptions about default domain IDs.

Reply:
What about restoring how it used to work in node options?

- Delete RMW_DEFAULT_DOMAIN_ID (or deprecate it).
- Add domain_id field to [rcl_init_options](https://github.com/ros2/rcl/blob/b9cfc243bf018091b5c5bcb5585d19591daf4ae6/rcl/src/rcl/init_options_impl.h#L28).
- The rcl API ([1](https://github.com/ros2/rcl/blob/b9cfc243bf018091b5c5bcb5585d19591daf4ae6/rcl/include/rcl/init_options.h#L172) and [2](https://github.com/ros2/rcl/blob/b9cfc243bf018091b5c5bcb5585d19591daf4ae6/rcl/include/rcl/init_options.h#L150)) should set and get the value in rcl, not directly the one in the rmw counterpart.
  - The default constructed rcl init options are set with `domain_id=RCL_DEFAULT_DOMAIN_ID`.
  - The default constructed rmw init options are set with `domain_id=0`.
  - If the user sets a domain id using [1](https://github.com/ros2/rcl/blob/b9cfc243bf018091b5c5bcb5585d19591daf4ae6/rcl/include/rcl/init_options.h#L172) and [2](https://github.com/ros2/rcl/blob/b9cfc243bf018091b5c5bcb5585d19591daf4ae6/rcl/include/rcl/init_options.h#L150), `rcl_init_options_t::domain_id` will be modified (and not its rmw counterpart).
  - When `rcl_init` is called, if `rcl_init_options_t::domain_id` is `RCL_DEFAULT_DOMAIN_ID` the environment is check and `rmw_init_options_t::domain_id` is set with the correct value. Otherwise, `rmw_init_options_t::domain_id` is a copy of `rcl_init_options_t::domain_id`.
  - `rcl_init_options_get_domain_id` will always return `rcl_init_options_t::domain_id` (this can return the unresolved default domain id).
  - the [node actual domain id](https://github.com/ros2/rcl/blob/b9cfc243bf018091b5c5bcb5585d19591daf4ae6/rcl/src/rcl/node.c#L59) is a copy of the `rmw_init_options::domain_id` that the node used (i.e.: it will always be a resolved number, and not SIZE_MAX).

That will match fairly well how this used to work before the node/participant mapping change (I think we overlooked this details in the middle of some pretty big PRs).

> Well, an rmw implementation is expected to provide a default domain ID via rmw_init_options_init(). From an API perspective, the sole value I see with having a sentinel ID is to defer picking an actual default (allowing, say, an implementation to pick up a default from a network-wide pool when it is actually needed and not when options are initialized).

Sure, but do we need that feature? I think this changes were a bit accidental and that the original purpose of the default domain id wasn't this.
I don't see a strong use case for letting the rmw implementation decide the domain id.

Reply:
> Sure, but do we need that feature?

That's a fair question that the @ros2/team has to answer. 

Personally, when I build a library, since I cannot predict (and likely won't see half of) all use cases, I try to keep it as general as possible unless there's a good reason not to. As opposed to when I build a particular system with a concrete known purpose for which I wouldn't add feature I know I don't need. But that's me.

Reply:
> Personally, when I build a library, since I cannot predict (and likely won't see half of) all use cases, I try to keep it as general as possible unless there's a good reason not to. As opposed to when I build a particular system with a concrete known purpose for which I wouldn't add a feature I know I don't need. But that's me.

Ok, we can add a `rmw_get_default_domain_id` and implement it in the three rmw implementations.
Then we can modify `rcl` to do something like:
```c
  domain_id = node->impl->options.domain_id;
  if (RCL_DEFAULT_DOMAIN_ID == domain_id) {
    if (RCL_RET_OK != rcl_get_default_domain_id(&domain_id)) {
      goto fail;
    }
+   if (RCL_DEFAULT_DOMAIN_ID == domain_id) {
+     domain_id = rmw_get_default_domain_id();
+   }
  }
  RCUTILS_LOG_DEBUG_NAMED(ROS_PACKAGE_NAME, "Using domain ID of '%zu'", domain_id);
```

In Foxy we would still need to set it to `0`, as we probably don't want to backport new API unnecessarily.
Does that sound good?

Reply:
Yeap. For a quick and dirty fix forcing to `0` is OK. For the rest, don't trust me. Let's wait for others to comment.

Reply:
For this point:

> There's absolutely no guarantee that rmw_init_options_init() will use RMW_DEFAULT_DOMAIN_ID as opposed to 42 or 5. Which would then break environment checks.

I think it must use `RMW_DEFAULT_DOMAIN_ID`. When the domain id is in the init options, that is just a default value, the user may override it. If the user wants the middleware to select the domain id, then `RMW_DEFAULT_DOMAIN_ID` should be used.

The purpose of `rmw_init_options_init()` was not to allow the middleware to specify a custom default domain id. Though it could have been that or still could be, though I don't think that makes sense... The user needs to be able to explicitly ask the middleware to pick a domain ID regardless of what is in the init options initially, and that is what the `RMW_DEFAULT_DOMAIN_ID` is about, from my perspective.

----

I think that we should remove the domain id from the node (node options, `actual_domain_id`, etc) and move it into the rmw context object. I think we should keep the `RMW_DEFAULT_DOMAIN_ID` sentinel value, and it can be set in the rmw init options (we could require middlewares initialize it to that value) and then it will always be resolved in the resulting rmw context object.

The reason for keeping the `RMW_DEFAULT_DOMAIN_ID`, in my opinion, is to let the domain id being used be configured in the middleware, maybe based on some middleware specific configuration file (similar to our `_DEFAULT` QoS settings we have) or as @hidmic said based on some dynamic heuristic like a pool of domain id's.

I do not think we need a function to query what the default domain ID would be if we initialized with `RMW_DEFAULT_DOMAIN_ID` (e.g. `rmw_get_default_domain_id()`), but I could be convinced otherwise. I think it would be sufficient to say you need to initialize the system in order to find out what the default is.

In `rcl`, if the environment variable `ROS_DOMAIN_ID` is set, then it should be honored, always. The only time the middleware would choose a domain id, is if the `ROS_DOMAIN_ID` is not set and the `RMW_DEFAULT_DOMAIN_ID` value is used, or if `ROS_DOMAIN_ID` is somehow set to the `RMW_DEFAULT_DOMAIN_ID` value.

I think `rcl_get_default_domain_id()` should be renamed probably, since it does not get the default, instead it gets the value from the environment variable if it is set otherwise it does nothing to the given `size_t *`.

Reply:
> I think that we should remove the domain id from the node (node options, actual_domain_id, etc) and move it into the rmw context object. I think we should keep the RMW_DEFAULT_DOMAIN_ID sentinel value, and it can be set in the rmw init options (we could require middlewares initialize it to that value) and then it will always be resolved in the resulting rmw context object.

:+1:

> The reason for keeping the RMW_DEFAULT_DOMAIN_ID, in my opinion, is to let the domain id being used be configured in the middleware, maybe based on some middleware specific configuration file (similar to our _DEFAULT QoS settings we have) or as @hidmic said based on some dynamic heuristic like a pool of domain id's.

:+1:

> I do not think we need a function to query what the default domain ID would be if we initialized with RMW_DEFAULT_DOMAIN_ID (e.g. rmw_get_default_domain_id()), but I could be convinced otherwise. I think it would be sufficient to say you need to initialize the system in order to find out what the default is.

We currently don't have any way to introspect the `domain_id` that the rmw implementation set, we would need either:
- rmw_context_get_domain_id(rmw_context_t * node)
- rmw_node_get_domain_id(rmw_node_t * node)

the first one makes more sense.
We can call that function after initing the `rmw_context` in `rcl`, and store it in the corresponding `rcl_context_t` structure.

Does that make sense?

Reply:
Why do we need a function in rmw to get that field? Could it not just be a public member of the `rmw_context_t` object? I mean either would work, I'm just wondering if there's a reason to hide this member internally or not.

Otherwise, lgtm.

----

One lingering doubt in my mind, should we leave the domain id in the node just so middlewares that choose to do it that way can still map domain id to nodes rather than contexts? I'm not sure it makes sense to do that, but I'm just asking if anyone thinks we'll need that at some point?

Reply:
> Why do we need a function in rmw to get that field? Could it not just be a public member of the rmw_context_t object? I mean either would work, I'm just wondering if there's a reason to hide this member internally or not.

Yeah, a member variable also sounds good.

> One lingering doubt in my mind, should we leave the domain id in the node just so middlewares that choose to do it that way can still map domain id to nodes rather than contexts? I'm not sure it makes sense to do that, but I'm just asking if anyone thinks we'll need that at some point?

My mental idea is that you have to create two contexts if you need different domain ids, regardless if the implementation can have a different domain id for each node or not.
Multi domain id applications are strange (mainly domain bridges?), and support for setting the domain id with node granularity doesn't seem to add much value.

Reply:
It's settled then! 

> My mental idea is that you have to create two contexts if you need different domain ids, regardless if the implementation can have a different domain id for each node or not.
Multi domain id applications are strange (mainly domain bridges?), and support for setting the domain id with node granularity doesn't seem to add much value.

I agree. I had my doubts about this at some point in time, but I see now those were unfounded.